src/opt/LowerISPCIntrinsics.cpp

```diff
@@ -0,0 +1,327 @@
+/*
+  Copyright (c) 2024, Intel Corporation
+
+  SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include "LowerISPCIntrinsics.h"
+
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/Transforms/Utils/Local.h>
+
+#include <numeric>
+
+namespace ispc {
+
+static llvm::Constant *lGetSequentialMask(llvm::IRBuilder<> &builder, unsigned N) {
+    std::vector<unsigned> indices(N);
+    std::iota(indices.begin(), indices.end(), 0);
+    return llvm::ConstantDataVector::get(builder.getContext(), indices);
+}
+
+static unsigned lGetVecNumElements(llvm::Value *V) {
+    auto *vecType = llvm::dyn_cast<llvm::VectorType>(V->getType());
+    Assert(vecType);
+    return vecType->getElementCount().getKnownMinValue();
+}
+
+static llvm::Value *lLowerConcatIntrinsic(llvm::CallInst *CI) {
+    llvm::IRBuilder<> builder(CI);
+    llvm::Value *V0 = CI->getArgOperand(0);
```

Comment:
Here and in similar places below please check that intrinsic has required number of arguments. I know that we often don't check it in optimizations but it's a big new change that will be a foundation for ISPC targets, so it should work perfectly.

Reply:
Done